keps/sig-scheduling/20190221-even-pods-spreading.md

```diff
@@ -0,0 +1,380 @@
+
+`EvenPodsSpreading` feature gives users more fine-grained control on
+distribution of pods scheduling, so as to achieve better high availability and
+resource utilization.
+
+## Motivation
+
+In Kubernetes, "Affinity" related directives are aimed to control how pods are
+scheduled - more packed or more scattering. But right now only limited options
+are offered: for `PodAffinity`, infinite pods can be stacked onto qualifying
+topology domain(s); for `PodAntiAffinity`, pods are scheduled onto each single
+topology domain exclusively.
+
+This is not an ideal situation if users want to put pods evenly across different
+topology domains - for the sake of high availability or saving cost. And regular
+rolling upgrade or scaling out replicas can also be problematic. See more
+details in [user stories](#user-stories).
+
+### Goals
+
+- Even spreading is calculated among pods instead of apps API (such as
+  Deployment, ReplicaSet).
+- Even spreading can be either a predicate (hard requirement) or a priority
+  (soft requirement).
+
+### Non-Goals
+
+- Even spreading is NOT calculated on an application basis. In other words, it's
+  not _only_ applied within replicas of an application.
+- "Max number of pods per topology" is NOT a goal.
+- Scale-down on an application is not guaranteed to achieve even pods spreading
+  in the initial implementation.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an application developer, I want my application pods to be scheduled onto
+specific topology domains as even as possible. Current status is that pods may
+be stacked onto a specific topology domain. (see
+[#68981](https://github.com/kubernetes/kubernetes/issues/68981))
+
+#### Story 2
+
+As an application developer, I want my application pods not to co-exist with
+specific pods (via PodAntiAffinity). But in some cases, it'd be favorable to
+tolerate "violating" pods in a manageable way. For example, suppose an app
+(replicas=2) is using PodAntiAffinity and deployed onto a 2-nodes cluster, and
+next the app needs to perform a rolling upgrade, then a third replacement pod is
+created, but it failed to be placed due to lack of resource. In this case,
+
+- if CA is enabled, a new machine will be provisioned to hold the new pod
+  (although old replicas will be deleted afterwards) (see
+  [#40358](https://github.com/kubernetes/kubernetes/issues/40358))
+- if CA is not enabled, it's a deadlock since the replacement pod can't be
+  placed. The only workaround at this moment is to update app strategyType from
+  "RollingUpdate" to "Recreate".
+
+Neither of them is an ideal solution. A promising solution is to give user an
+option to trigger "toleration" mode when the cluster is out of resource. Then in
+aforementioned example, a third pod is "tolerated" to be put onto node1 (or
+node2). But keep it in mind, this behavior is only triggered upon resource
+shortage. For a 3-nodes cluster, the third pod will still be placed onto node3
+(if node3 is capable).
```

Comment:
For podantiaffinity, we should trigger the toleration mode in other scenarios not just resource shortage. I mean, podantiaffinity could be specified in hard or soft manner. One which says i cannot absolutely tolerate A to coexist with B, or as much as possible keep A away from B. So in strict case, i will only have a zone contain one of A or B. But in relaxed case, i will have  A distribution as 2/1/1 and as new pods of B arrive, they are placed on zones with less of A, so they are distributed as 0/1/0--->0/1/1 ---> 0/2/1 or 0/1/2  . Basically we would need to define when the new disitribution would increase the overall separation between A and B per zone more. I know this is hard to reason about :-)

Reply:
In a recent discussion, we gave up the design to build things on top of PodAntiAffinity (which is way complicated). Instead, build a separate feature which focuses on "even pods spreading", and it can work with any other predicates/priorities.

So in new design, in terms of PodAntiAffinity:

- if you use `PodAntiAffinity`, it's the same as before - only one pod can be scheduled onto one topology domain
- if you want the "toleration mode" as above described, don't use `PodAntiAffinity`, use the new feature with proper fields (maxSkew, topologyKey and "podSelector". PS: API for "podSelector" hasn't finalized yet)

Reply:
ok makes sense. So the new feature is only about spreading with some skewness. we need this documented that anti affinity needs to be using the old feature . That also means we cannot deprecate anti affinity or how show using an example how to achieve it using  new feature